[[ Bug 19335 ]] Add sized integer types to LCB

runrevmark opened the pull request:
This patch adds the complete set of integer types both explicitly sized and C types. Furthermore it makes Java primitive type mappings more strict.

runrevmark committed:
libfoundation: Add MCHashBool

This patch adds an MCHashBool function which hashes a C bool
value.

runrevmark committed:
libfoundation: Rework foreign value typeinfo creation

This patch reworks the libfoundation foreign type creation to use
templates and remove duplicate code wherever possible.

runrevmark committed:
libfoundation: Add sized integer types to foundation-foreign

This patch adds signed and unsigned sized integer types as foreign
values.

peter-b commented on libfoundation/src/foundation-foreign.cpp:
```diff
@@ -182,545 +321,403 @@ bool __MCForeignValueCopyDescription(__MCForeignValue *self, MCStringRef& r_desc
-{
-    return __any_int_export<size_t>(value, release, contents, kMCSizeTypeInfo);
-}
+    static bool doexport(MCNumberRef p_boxed_value, typename RealType::c_type& r_value)
```
It might be worth extracting the C type as a template parameter, just in order to avoid repeating `typename RealType::c_type`.  Don't forget we're allowed default template parameters now!

```c++
template <typename TypeDesc, typename RealType = typename TypeDesc::c_type>
struct DoExport<TypeDesc,
    typename std::enable_if<std::is_floating_point<RealType>::value>::type>
{
    static bool doexport(MCNumberRef p_boxed_value, RealType& r_value)
    {
        r_value = RealType(MCNumberFetchAsReal(p_boxed_value));
        return true;
    }
}
```

I guess you could apply this pattern in several places where you're currently tediously extracting the `c_type` from the type descriptor struct.

runrevmark replied:
I tried that - but it doesn't seem to work as this is a partial specialization :(

peter-b reviewed (commented):
Looking good to me

runrevmark committed:
libfoundation: Add sized integer types to foundation-foreign

This patch adds the C long / unsigned long type alongside the other
sized integer types and additionally adds checks for the size of
int and uint.

runrevmark committed:
libscript: Clean up foreign module integer types

This patch cleans up the foreign module integer types, adding
the full set of sized integer types, and the full set of C integer
types.

Most C types alias the appropriate sized integer type, but a
new pair for unsigned long and long have been added as foreign
types.

runrevmark committed:
libfoundation: Fix incorrect check for size of long in 64-bit builds

This patch corrects the incorrect static_asser clause in clong_type_desc_t

runrevmark committed:
libfoundation: Fix pointer foreign type defined operation

This patch corrects the generic implementation of 'defined' - ensuring
it returns true if the value of the slot is not the default value
of the type.

runrevmark committed:
libfoundation: Make java ffi map int types correctly

This patch ensures that java uses appropriate mappings to the
sized integer types.

Additionally it adds J* type definitions to the java lcb module
to make foreign methods types clearer to express.

runrevmark committed:
libfoundation: Add unit tests for foreign value types

This patch adds C++ tests for the foreign value types defined
by libfoundation.

runrevmark committed:
libfoundation: Fix issues highlighted by unit tests

This patch fixes a number of issues highlighted by the unit tests:

    - the describe method was not being copied into the typeinfo
    - the wrong error typeinfo was being used for integral imports
    - the format strings were inconsistent (use 'd' instead of 'i')
    - the real format string has been changed to 'g'
    - the correct hash function is being used for size types

runrevmark committed:
lcb: Add release note and docs for sized integer types

This patch adds a release note and modifies the LCB language
reference to include the newly added sized integer types.

runrevmark committed:
[[ Tests ]] Fix an error in TestLoadExtension

This patch ensures that an appropriate error is generated if
the extension is not found.

runrevmark committed:
lcb: Integer type improvements

This patch reworks the sized integer types ensuring that all types
which may be distinct are truly distinct. In doing so it unifies
the naming of all types (S prefix for signed integers, U for unsigned
integers), adds all the C integral types, and adds aliases from
Int* to SInt*.

A particular outcome of this patch is that foreign handlers which
bind to Java methods *must* use the J* types and not the C* types
as the C types are now considered distinct, and the Java binding
code expects the J types (which are just aliases to the machine
types since the JNI fixes that mapping).

peter-b commented on libfoundation/src/foundation-foreign.cpp:
```diff
@@ -25,83 +25,176 @@
-    static constexpr auto& bridge_type_info = kMCBooleanTypeInfo;
-    static constexpr auto& hash_func = MCHashBool;
+template<typename CType>
+constexpr MCForeignPrimitiveType __compute_signed_primitive_type()
```
I suggest avoiding adding new symbols that contain `__`.  I think it would be okay to name this `compute_signed_primitive_type()`.

peter-b commented on libfoundation/src/foundation-foreign.cpp:
```diff
@@ -25,83 +25,176 @@
+    return kMCForeignPrimitiveTypeVoid;
+}
+template<> constexpr MCForeignPrimitiveType
+__compute_primitive_type<char>(void) { return __compute_signed_primitive_type<char>(); }
```
Unfortunately you are assuming that `char` is signed.

> The signedness of char depends on the compiler and the target platform: the defaults for ARM and PowerPC are typically unsigned, the defaults for x86 and x64 are typically signed.

I think it might be a good idea to explore the possibility of having two templates for `compute_primitive_type<T>()`, which use `enable_if<is_integral<T>::value && is_unsigned<T>::value>::type` to enable the correct one.  If that was possible, it would solve the "is `char` signed?" problem, and also might allow many of the explicit `compute_primitive_type<T>()` specialisations to be deleted.

runrevmark replied:
It isn't assuming char is signed - it is just using the signed_primitive operation to compute a suitable primitive type based on its size. libffi has no mapping for 'char' - but it can be treated as either a signed or unsigned integer quantity from its point of view.

runrevmark replied:
I must confess my C++-fu was not quite up to figuring out how to do the enable_if stuff, hence this approach... Any suggestions welcome!

peter-b replied:
<details><summary>Here's one possible approach</summary>

```c++

template <typename CType, typename Enable = void>
struct compute_primitive_type
{
    static_assert(sizeof(CType) == 0,
                  "No mapping from CType to primitive type");
};

template <typename CType>
struct compute_primitive_type<
    CType, typename std::enable_if<std::is_integral<CType>::value &&
                                   std::is_signed<CType>::value>::type>
{
    static_assert(sizeof(CType) != 1 || sizeof(CType) != 2 ||
                  sizeof(CType) != 4 || sizeof(CType) != 8,
                  "Unsupported signed integer size");
    static constexpr auto value =
        sizeof(CType) == 1 ? kMCForeignPrimitiveTypeSInt8 :
        sizeof(CType) == 2 ? kMCForeignPrimitiveTypeSInt16 :
        sizeof(CType) == 4 ? kMCForeignPrimitiveTypeSInt32 :
        sizeof(CType) == 8 ? kMCForeignPrimitiveTypeSInt64 :
        kMCForeignPrimitiveTypeVoid;
};

template <typename CType>
struct compute_primitive_type<
    CType, typename std::enable_if<std::is_integral<CType>::value &&
                                   !std::is_signed<CType>::value>::type>
{
    static_assert(sizeof(CType) != 1 || sizeof(CType) != 2 ||
                  sizeof(CType) != 4 || sizeof(CType) != 8,
                  "Unsupported unsigned integer size");
    static constexpr auto value =
        sizeof(CType) == 1 ? kMCForeignPrimitiveTypeUInt8 :
        sizeof(CType) == 2 ? kMCForeignPrimitiveTypeUInt16 :
        sizeof(CType) == 4 ? kMCForeignPrimitiveTypeUInt32 :
        sizeof(CType) == 8 ? kMCForeignPrimitiveTypeUInt64 :
        kMCForeignPrimitiveTypeVoid;
};

template <> struct compute_primitive_type<float>
{
    static constexpr auto value = kMCForeignPrimitiveTypeFloat32;
};

template <> struct compute_primitive_type<double>
{
    static constexpr auto value = kMCForeignPrimitiveTypeFloat64;
};

template<typename CType>
struct numeric_type_desc_t {
    using c_type = CType;
    static constexpr auto primitive_type = compute_primitive_type<CType>::value;
    static constexpr auto& base_type_info = kMCNullTypeInfo;
    static constexpr auto is_optional = false;
    static constexpr auto is_bridgable = true;
    using bridge_type = MCNumberRef;
    static constexpr auto& bridge_type_info = kMCNumberTypeInfo;
};
```

</details>

peter-b replied:
I tried a `constexpr` template function approach to the same thing, but I get "non-type partial specialization ... is not allowed" compile errors.  I think this approach is reasonable because it has the same `magic_struct<T>::value` pattern that similar templates in the STL use.

peter-b commented on libfoundation/src/foundation-foreign.cpp:
```diff
@@ -238,6 +280,92 @@ struct pointer_type_desc_t {
+};
+struct cchar_type_desc_t: public integral_type_desc_t<char> {
+    static constexpr auto& type_info = kMCCCharTypeInfo;
+    static constexpr auto describe_format = "<foreign c char '%c'>";
```
Do you think it would be reasonable to capitalize "C" when it refers to the programming language?

runrevmark replied:
Hmm - I pondered that - the rationale for using 'c' is that in LCB, if we make foreign binding more explicit then:
```
  foreign c handler ...
```
Might be more consistent (with regards token casing rules) as 'c' and not 'C'.

runrevmark committed:
lcb: Improve implementation of 'compute_primitive_type'

This patch uses more economic templates with 'enable_if' to
implement the compute_primitive_type() function which maps
a CType to the MCForeignPrimitiveType enum member it maps to.

livecodeali commented on tests/lcb/stdlib/java.lcb:
```diff
@@ -40,7 +40,7 @@ end handler
 foreign handler CreateJavaString(in pBytes as JByteArray) returns JString binds to "java:java.lang.String>new([B)"
 
-foreign handler MCDataCreateWithBytes(in pBytes as ZStringNative, in pCount as CUInt, out rData as Data) returns CBool binds to "<builtin>"
+foreign handler MCDataCreateWithBytes(in pBytes as ZStringNative, in pCount as CUInt, out rData as Data) returns JBoolean binds to "<builtin>"
```
This one's binding to the libfoundation function so should remain `CBool` shouldn't it?

livecodeali commented on docs/guides/LiveCode Builder Language Reference.md:
```diff
@@ -430,31 +430,48 @@ statement blocks.
+ - JByte maps to Int8
+ - JShort maps to Int16
+ - JInt maps to Int32
+ - JLong maps to Int64
```
This list is missing JFloat and JDouble

livecodeali commented on docs/lcb/notes/feature-sized_ints.md:
```diff
@@ -0,0 +1,17 @@
+  These all map to their corresponding C counterparts.
+
+* The Java integer primtive types JByte, JShort, JInt, JLong,
+  JFloat and JDouble have been added. These map to int8_t,
```
and `JBoolean`

livecodeali commented on libfoundation/src/foundation-foreign.cpp:
```diff
@@ -182,559 +496,487 @@ bool __MCForeignValueCopyDescription(__MCForeignValue *self, MCStringRef& r_desc
+    static_assert(TypeDesc::is_bridgable, "This type is not bridgable");
+    bool t_success = DoImport<TypeDesc>::doimport(*static_cast<typename TypeDesc::c_type *>(contents),
+                                        reinterpret_cast<typename TypeDesc::bridge_type&>(r_value));
+    return t_success;
```
not a big deal, but why doesn't this just `return`?

runrevmark committed:
lcb: Minor cleanup

This patch cleans up a number of minor things in the patch. Including:

  - ensuring the java type definitions use machine types, and not C
    types

  - using the correct int for the bind to MCDataCreateWithBytes

  - adding some missing new types to the docs

livecodeali commented:
@livecode-vulcan review ok c9e8447

livecode-vulcan commented:
💙 review by @livecodeali ok c9e84471092f65ea07ec66247564b0a623a5d592

livecode-vulcan commented:
😞 test failure c9e84471092f65ea07ec66247564b0a623a5d592

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/44)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/43)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/30)
* try-community-universal-ios-iphoneos10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.2/builds/48)
* try-community-universal-ios-iphonesimulator10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.2/builds/48)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/42)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/44)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/46)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/48)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/46)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/48)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/48)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/45)

runrevmark commented:
@livecode-vulcan review ok b9b47a5

runrevmark committed:
lcb: Change definition of 'type_info' members in foreign templates

This patch works-around a deficiency in VS2015 where:

  extern MCTypeInfoRef MyFoo;
  constexpr auto& foo = MyFoo;

Will not compile. Instead, constexpr functions are now used:

  constexpr MCTypeInfoRef& foo() { return MyFoo; }

livecode-vulcan commented:
💙 review by @runrevmark ok b9b47a55f63cc59fb510173c224c931db84dd941

livecode-vulcan commented:
😞 test failure b9b47a55f63cc59fb510173c224c931db84dd941

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/47)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/46)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/33)
* try-community-universal-ios-iphoneos10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.2/builds/51)
* try-community-universal-ios-iphonesimulator10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.2/builds/51)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/45)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/47)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/49)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/51)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/49)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/51)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/51)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/48)

runrevmark committed:
lcb: Fix 32-bit architecture errors in sized integer tests

This patch fixes a number of errors when running on 32-bit processors
in the sized integer tests.

Additionally, it adds MCHashInt64 and MCHashUInt64 as they are not
covered by any other MCHash function on 32-bit architectures.

runrevmark commented:
Note: Self-review to check that everything now passes!

@livecode-vulcan a8c896b

runrevmark commented:
@livecode-vulcan review ok a8c896b

livecode-vulcan commented:
💙 review by @runrevmark ok a8c896b711c158302350c6c7d06a5a73d875c874

livecode-vulcan commented:
😎 test success a8c896b711c158302350c6c7d06a5a73d875c874

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/52)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/51)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/38)
* try-community-universal-ios-iphoneos10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.2/builds/56)
* try-community-universal-ios-iphonesimulator10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.2/builds/56)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/50)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/52)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/54)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/56)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/54)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/56)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/56)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/53)

livecodeali commented:
@runrevmark I've just looked over the changes, you can consider this officially reviewed :-)
